File: src/main/java/com/kishlaly/ta/cache/IndicatorsInMemoryCache.java
package com.kishlaly.ta.cache;

import com.google.gson.Gson;
import com.kishlaly.ta.cache.key.*;
import com.kishlaly.ta.model.Timeframe;
import com.kishlaly.ta.model.indicators.*;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

public class IndicatorsInMemoryCache {

    private static Gson gson = new Gson();

    private static ConcurrentHashMap<EMAKey, List<EMAJava>> ema = new ConcurrentHashMap<>();
    private static ConcurrentHashMap<MACDKey, List<MACD>> macd = new ConcurrentHashMap<>();
    private static ConcurrentHashMap<KeltnerKEY, List<KeltnerJava>> keltner = new ConcurrentHashMap<>();
    private static ConcurrentHashMap<ATRKey, List<ATRJava>> atr = new ConcurrentHashMap<>();
    private static ConcurrentHashMap<StochKey, List<Stoch>> stoch = new ConcurrentHashMap<>();
    private static ConcurrentHashMap<BollingerKey, List<BollingerJava>> bollinger = new ConcurrentHashMap<>();
    private static ConcurrentHashMap<EFIKey, List<ElderForceIndexJava>> efi = new ConcurrentHashMap<>();

    public static void putEMA(String symbol, Timeframe timeframe, int period, List<EMAJava> data) {
        ema.put(new EMAKey(symbol, timeframe, period), data);
    }

    public static void putMACD(String symbol, Timeframe timeframe, List<MACD> data) {
        macd.put(new MACDKey(symbol, timeframe), data);
    }

    public static void putKeltner(String symbol, Timeframe timeframe, List<KeltnerJava> data) {
        keltner.put(new KeltnerKEY(symbol, timeframe), data);
    }

    public static void putATR(String symbol, Timeframe timeframe, int period, List<ATRJava> data) {
        atr.put(new ATRKey(symbol, timeframe, period), data);
    }

    public static void putStoch(String symbol, Timeframe timeframe, List<Stoch> data) {
        stoch.put(new StochKey(symbol, timeframe), data);
    }

    public static void putEFI(String symbol, Timeframe timeframe, List<ElderForceIndexJava> data) {
        efi.put(new EFIKey(symbol, timeframe), data);
    }

    public static void putBollinger(String symbol, Timeframe timeframe, List<BollingerJava> data) {
        bollinger.put(new BollingerKey(symbol, timeframe), data);
    }

    public static List<EMAJava> getEMA(String symbol, Timeframe timeframe, int period) {
        List<EMAJava> cached = ema.getOrDefault(new EMAKey(symbol, timeframe, period), Collections.emptyList());
        if (!cached.isEmpty()) {
            String json = gson.toJson(cached);
            List<EMAJava> copy = gson.fromJson(json, new com.google.common.reflect.TypeToken<List<EMAJava>>() {
            }.getType());
            Collections.sort(copy, Comparator.comparing(EMAJava::getTimestamp));
            return copy;
        }
        return Collections.emptyList();
    }


    public static List<MACD> getMACD(String symbol, Timeframe timeframe) {
        List<MACD> cached = macd.getOrDefault(new MACDKey(symbol, timeframe), Collections.emptyList());
        if (!cached.isEmpty()) {
            String json = gson.toJson(cached);
            List<MACD> copy = gson.fromJson(json, new com.google.common.reflect.TypeToken<List<MACD>>() {
            }.getType());
            Collections.sort(copy, Comparator.comparing(MACD::getTimestamp));
            return copy;
        }
        return Collections.emptyList();
    }


    public static List<KeltnerJava> getKeltner(String symbol, Timeframe timeframe) {
        List<KeltnerJava> cached = keltner.getOrDefault(new KeltnerKEY(symbol, timeframe), Collections.emptyList());
        if (!cached.isEmpty()) {
            String json = gson.toJson(cached);
            List<KeltnerJava> copy = gson.fromJson(json, new com.google.common.reflect.TypeToken<List<KeltnerJava>>() {
            }.getType());
            Collections.sort(copy, Comparator.comparing(KeltnerJava::getTimestamp));
            return copy;
        }
        return Collections.emptyList();
    }


    public static List<ATRJava> getATR(String symbol, Timeframe timeframe, int period) {
        List<ATRJava> cached = atr.getOrDefault(new ATRKey(symbol, timeframe, period), Collections.emptyList());
        if (!cached.isEmpty()) {
            String json = gson.toJson(cached);
            List<ATRJava> copy = gson.fromJson(json, new com.google.common.reflect.TypeToken<List<ATRJava>>() {
            }.getType());
            Collections.sort(copy, Comparator.comparing(ATRJava::getTimestamp));
            return copy;
        }
        return Collections.emptyList();
    }


    public static List<Stoch> getStoch(String symbol, Timeframe timeframe) {
        List<Stoch> cached = stoch.getOrDefault(new StochKey(symbol, timeframe), Collections.emptyList());
        if (!cached.isEmpty()) {
            String json = gson.toJson(cached);
            List<Stoch> copy = gson.fromJson(json, new com.google.common.reflect.TypeToken<List<Stoch>>() {
            }.getType());
            Collections.sort(copy, Comparator.comparing(Stoch::getTimestamp));
            return copy;
        }
        return Collections.emptyList();
    }

    public static List<BollingerJava> getBollinger(String symbol, Timeframe timeframe) {
        List<BollingerJava> cached = bollinger.getOrDefault(new BollingerKey(symbol, timeframe), Collections.emptyList());
        if (!cached.isEmpty()) {
            String json = gson.toJson(cached);
            List<BollingerJava> copy = gson.fromJson(json, new com.google.common.reflect.TypeToken<List<BollingerJava>>() {
            }.getType());
            Collections.sort(copy, Comparator.comparing(BollingerJava::getTimestamp));
            return copy;
        }
        return Collections.emptyList();
    }

    public static List<ElderForceIndexJava> getEFI(String symbol, Timeframe timeframe) {
        List<ElderForceIndexJava> cached = efi.getOrDefault(new EFIKey(symbol, timeframe), Collections.emptyList());
        if (!cached.isEmpty()) {
            String json = gson.toJson(cached);
            List<ElderForceIndexJava> copy = gson.fromJson(json, new com.google.common.reflect.TypeToken<List<ElderForceIndexJava>>() {
            }.getType());
            Collections.sort(copy, Comparator.comparing(ElderForceIndexJava::getTimestamp));
            return copy;
        }
        return Collections.emptyList();
    }


    public static void clear() {
        ema.clear();
        macd.clear();
        keltner.clear();
        atr.clear();
        stoch.clear();
    }

}

File: src/main/kotlin/com/kishlaly/ta/cache/IndicatorsInMemoryCache.kt
package com.kishlaly.ta.cache

import com.google.common.reflect.TypeToken
import com.google.gson.Gson
import com.google.gson.internal.LinkedTreeMap
import com.kishlaly.ta.cache.keys.*
import com.kishlaly.ta.model.AbstractModel
import com.kishlaly.ta.model.Timeframe
import com.kishlaly.ta.model.indicators.*
import java.util.concurrent.ConcurrentHashMap

class IndicatorsInMemoryCache {

    companion object {
        private val gson: Gson = Gson()
        private val ema = ConcurrentHashMap<EMAKey, List<EMA>>()
        private val macd = ConcurrentHashMap<MACDKey, List<MACD>>()
        private val keltner = ConcurrentHashMap<KeltnerKey, List<Keltner>>()
        private val atr = ConcurrentHashMap<ATRKey, List<ATR>>()
        private val stochastic = ConcurrentHashMap<StochKey, List<Stochastic>>()
        private val bollinger = ConcurrentHashMap<BollingerKey, List<Bollinger>>()
        private val efi = ConcurrentHashMap<EFIKey, List<ElderForceIndex>>()

        fun putEMA(symbol: String, timeframe: Timeframe, period: Int, data: List<EMA>) {
            ema[EMAKey(symbol, timeframe, period)] = data
        }

        fun getEMA(symbol: String, timeframe: Timeframe, period: Int): List<EMA> {
            val get = get(ema, EMAKey(symbol, timeframe, period)) as ArrayList<LinkedTreeMap<Any, Any>>
            return get.map { EMA((it["timestamp"] as Double).toLong(), it["value"] as Double) }
                .toList().sortedBy { it.timestamp }
        }

        fun putMACD(symbol: String, timeframe: Timeframe, data: List<MACD>) {
            macd[MACDKey(symbol, timeframe)] = data
        }

        fun getMACD(symbol: String, timeframe: Timeframe): List<MACD> {
            val get = get(macd, MACDKey(symbol, timeframe)) as ArrayList<LinkedTreeMap<Any, Any>>
            return get.map {
                MACD(
                    (it["timestamp"] as Double).toLong(), it["macd"] as Double, it["signal"] as Double,
                    it["histogram"] as Double
                )
            }.toList().sortedBy { it.timestamp }
        }

        fun putKeltner(symbol: String, timeframe: Timeframe, data: List<Keltner>) {
            keltner[KeltnerKey(symbol, timeframe)] = data
        }

        fun getKeltner(symbol: String, timeframe: Timeframe): List<Keltner> {
            val get = get(keltner, KeltnerKey(symbol, timeframe)) as ArrayList<LinkedTreeMap<Any, Any>>
            val sortedBy = get.map {
                Keltner(
                    (it["timestamp"] as Double).toLong(),
                    it["low"] as Double,
                    it["middle"] as Double,
                    it["top"] as Double
                )
            }.toList().sortedBy { it.timestamp }
            return sortedBy
        }

        fun putATR(symbol: String, timeframe: Timeframe, period: Int, data: List<ATR>) {
            atr[ATRKey(symbol, timeframe, period)] = data
        }

        fun getATR(symbol: String, timeframe: Timeframe, period: Int): List<ATR> {
            val get = get(atr, ATRKey(symbol, timeframe, period)) as ArrayList<LinkedTreeMap<Any, Any>>
            return get.map { ATR((it["timestamp"] as Double).toLong(), it["value"] as Double) }
                .toList().sortedBy { it.timestamp }
        }

        fun putStoch(symbol: String, timeframe: Timeframe, data: List<Stochastic>) {
            stochastic[StochKey(symbol, timeframe)] = data
        }

        fun getStoch(symbol: String, timeframe: Timeframe): List<Stochastic> {
            val get = get(stochastic, StochKey(symbol, timeframe)) as ArrayList<LinkedTreeMap<Any, Any>>
            return get.map {
                Stochastic(
                    (it["timestamp"] as Double).toLong(),
                    it["slowD"] as Double,
                    it["slowK"] as Double
                )
            }.toList().sortedBy { it.timestamp }
        }

        fun putEFI(symbol: String, timeframe: Timeframe, data: List<ElderForceIndex>) {
            efi[EFIKey(symbol, timeframe)] = data
        }

        fun getEFI(symbol: String, timeframe: Timeframe): List<ElderForceIndex> {
            val get = get(efi, EFIKey(symbol, timeframe)) as ArrayList<LinkedTreeMap<Any, Any>>
            return get.map {
                ElderForceIndex(
                    (it["timestamp"] as Double).toLong(),
                    it["value"] as Double
                )
            }.toList().sortedBy { it.timestamp }
        }

        fun putBollinger(symbol: String, timeframe: Timeframe, data: List<Bollinger>) {
            bollinger[BollingerKey(symbol, timeframe)] = data
        }

        fun getBollinger(symbol: String, timeframe: Timeframe): List<Bollinger> {
            val get = get(bollinger, BollingerKey(symbol, timeframe)) as ArrayList<LinkedTreeMap<Any, Any>>
            return get.map {
                Bollinger(
                    (it["timestamp"] as Double).toLong(),
                    it["bottom"] as Double,
                    it["middle"] as Double,
                    it["top"] as Double
                )
            }.toList().sortedBy { it.timestamp }
        }

        private fun <T : AbstractModel, K : BaseKey> get(map: ConcurrentHashMap<K, List<T>>, key: K): List<T> {
            return copy(map.getOrDefault(key, emptyList()))
        }

        private fun <T : AbstractModel> copy(source: List<T>): List<T> {
            return gson.fromJson<List<T>>(gson.toJson(source), object : TypeToken<List<T>>() {}.type)
        }

        fun clear() {
            ema.clear()
            macd.clear()
            keltner.clear()
            atr.clear()
            stochastic.clear()
        }

    }

}